text/0000-ember-data-singleton-record-data.md

```diff
@@ -0,0 +1,450 @@
+    identifier: RecordIdentifier,
+    key: string
+  ): boolean {}
+}
+```
+
+</td>
+</tr>
+
+</tbody>
+</table>
+
+**Updates to RecordData Methods**
+
+<table>
+<thead>
+<tr>
+<th>Old</th>
+<th>New</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>
+
+```ts
+interface LegacyResourceIdentifierObject {
+  type: string;
+  id: string | null;
+  clientId: string | null;
+}
+
+interface RecordDataV1 {
+  unloadRecord(): void;
+  isRecordInUse(): boolean;
+  isEmpty(): boolean;
+  isNew(): boolean;
+
+  getAttr(propertyName: string): any;
+  isAttrDirty(propertyName: string): boolean;
+  changedAttributes(): AttributesChanges;
+  hasChangedAttributes(): boolean;
+  rollbackAttributes(): void;
+
+  @deprecated // this RFC deprecates this method entirely
+  getBelongsTo(propertyName: string): { data: LegacyResourceIdentifierObject };
+  @deprecated // this RFC deprecates this method entirely
+  getHasMany(propertyName: string): { data: LegacyResourceIdentifierObject[] };
+
+  willCommit(): void;
+  didCommit(data: any): void;
+
+  @deprecated // this RFC deprecates this method entirely
+  _initRecordCreateOptions(options: object): object;
+  clientDidCreate(): void;
+
+  @deprecated // this RFC deprecates this method entirely and without replacement
+  getResourceIdentifier(): LegacyResourceIdentifierObject;
+
+  // original RecordData RFC specified setBelongsTo
+  // with the 2nd arg being LegacyResourceIdentifierObject
+  // but that was not what the resulting implementation in
+  // ember-data did
+  setDirtyBelongsTo(propertyName: string, value: RecordData | null): void;
+
+  // original RecordData RFC specified setAttribute
+  // but that was not what the resulting implementation in
+  // ember-data did
+  setDirtyAttribute(propertyName: string, value: any): void;
+
+  // original RecordData RFC specified setHasMany
+  // with the 2nd arg being LegacyResourceIdentifierObject[]
+  // but that was not what the resulting implementation in
+  // ember-data did
+  setDirtyHasMany(propertyName: string, value: RecordData[]): void;
+
+  // original RecordData RFC specified addToHasMany
+  // with the 2nd arg being LegacyResourceIdentifierObject[]
+  // but that was not what the resulting implementation in
+  // ember-data did
+  addToHasMany(propertyName: string, values: RecordData[], startIndex?: number): void;
+
+  // original RecordData RFC specified removeFromHasMany
+  // with the 2nd arg being LegacyResourceIdentifierObject[]
+  // but that was not what the resulting implementation in
+  // ember-data did
+  removeFromHasMany(propertyName: string, values: RecordData[]): void;
+
+  @deprecated // this RFC deprecates this method entirely and without replacement
+  removeFromInverseRelationships(isNew: boolean): void;
+}
+```
+
+</td>
+<td>
+
+```ts
+interface RecordDataV2 {
+  unloadRecord(identifier: Identifier);
+  isRecordInUse(identifier: Identifier): boolean;
+  isEmpty(identifier: Identifier): boolean;
+  isNew(identifier: Identifier): boolean;
```

Comment:
Why do we have record in `isRecordInUse` but not in `isNew`?  No doubt it's a holdover from the original state flags, but for a new api does this naming distinction make sense?

Reply:
None of the updates for RecordData state use `Record` in the names, so at first I was tempted to say "it doesn't make sense" and "we should drop it".

https://github.com/igort/rfcs/blob/igor/record-state/text/0000-record-data-state.md

But then I decided to investigate a bit more about why `isRecordInUse` exists:

`RecordData.isRecordInUse` calls `StoreWrapper.isRecordInUse` which calls `InternalModel.isRecordInUse`.

It is only used in one place (by `_cleanupOrphanedRecordDatas`)

And it was unspecified by the original RFC for both the StoreWrapper and the RecordData location: https://github.com/emberjs/rfcs/blob/master/text/0293-record-data.md

I think we should take this opportunity to drop it entirely as an internal concern. It essentially was something we leaked due to the relationship layer using RecordData instances as identifiers, but once we refactor that layer to use identifiers instead the need for it here dies.

Reply:
That sounds good to me.  We should make sure this behaviour is tested of course; but I agree that the "unload" story with relationships is quite specific to the relationship layer code.